File: Tree/BinaryTree/max_depth.cpp
#include <iostream>
#include <algorithm> // for max()
using namespace std;

class TreeNode {
public:
    int data;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int data) {
        this->data = data;
        left = nullptr;
        right = nullptr;
    }
};

/*
Function: maxDepth()
Purpose : Returns the maximum depth (or height) of the binary tree.
          Depth = number of nodes along the longest path from root to leaf.

Approach:
    - Use recursion.
    - For each node:
        → compute the depth of left subtree (lh)
        → compute the depth of right subtree (rh)
        → answer = 1 + max(lh, rh)

Time Complexity:
    O(N) — every node is visited once.

Space Complexity:
    O(H) — recursive call stack, where H is height of tree.
           Worst case (skewed tree): O(N)
           Best case (balanced tree): O(log N)
*/
int maxDepth(TreeNode* root) {
    if (root == nullptr) return 0;

    int lh = maxDepth(root->left);
    int rh = maxDepth(root->right);

    return 1 + max(lh, rh);
}

int main() {
    // Example tree:
    //        1
    //       / \
    //      2   3
    //     / \
    //    4   5

    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);

    cout << "Maximum Depth of Tree: " << maxDepth(root) << endl;
    return 0;
}

File: java/Tree/BinaryTree/MaxDepth.java
package Tree.BinaryTree;

/**
 * Tree Node Definition
 */
class TreeNode {
    int data;
    TreeNode left;
    TreeNode right;

    TreeNode(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}

public class MaxDepth{

    /*
     * Function: maxDepth()
     * Purpose : Returns the maximum depth (height) of the binary tree.
     *           Depth = number of nodes along the longest path
     *           from root to leaf.
     *
     * Approach:
     *  - Use recursion.
     *  - For each node:
     *      → compute left subtree height
     *      → compute right subtree height
     *      → result = 1 + max(left, right)
     *
     * Time Complexity:
     *  O(N) — each node is visited once.
     *
     * Space Complexity:
     *  O(H) — recursive call stack
     *      Worst case (skewed tree): O(N)
     *      Best case (balanced tree): O(log N)
     */
    static int maxDepth(TreeNode root) {
        if (root == null)
            return 0;

        int leftHeight = maxDepth(root.left);
        int rightHeight = maxDepth(root.right);

        return 1 + Math.max(leftHeight, rightHeight);
    }

    public static void main(String[] args) {

        // Example tree:
        //        1
        //       / \
        //      2   3
        //     / \
        //    4   5

        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        System.out.println("Maximum Depth of Tree: " + maxDepth(root));
    }
}
